### PR TITLE
Add tab completion for run command

### PR DESCRIPTION
**Pull request by cgranleese-r7**
Re-opening https://github.com/rapid7/metasploit-framework/pull/14150

### Before
Tab completion was not compatible with the run command.

### After
Tab completion now works with relevant modules:
- Auxiliary
- Evasion
- Exploit
- Payload
- Post

add_tab_completion_pr

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use exploit/windows/wins/ms04_045_wins` - just an example, possibly test different module types
- [ ] type `run` followed by a space then hit `tab`
- [ ] **Verify** that a list of possible tab completions displays
- [ ] **Verify** that tab completion works for both flags and options in multiple different scenarios 

#### Completing keys:

```
use smb_version
run rhos<tab>
```

Expected result: `run rhosts=`

#### Completing values:

```
use smb_version
run rhosts=<tab>
```

Expected result: `run rhosts=x.x.x.x `

### Completing keys, and unrelated commands

```
use smb_version
run rhosts<tab>
db_statu<tab>
```

Expected result: _with a space_: `db_status `.

### CODE REVIEW

File: .rakeTasks [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (dwelch-r7) [inline .rakeTasks:1]: was this added by accident?

File: lib/rex/ui/text/dispatcher_shell.rb [hunk 2]
```diff
@@ -5,661 +5,650 @@
...
+          #
+          def tab_complete_filenames(str, words)
+            matches = ::Readline::FILENAME_COMPLETION_PROC.call(str)
+            if matches and matches.length == 1 and File.directory?(matches[0])
+              dir = matches[0]
+              dir += File::SEPARATOR if dir[-1,1] != File::SEPARATOR
+              matches = ::Readline::FILENAME_COMPLETION_PROC.call(dir)
+            end
+            matches
+          end
```
Comment (dwelch-r7) [inline lib/rex/ui/text/dispatcher_shell.rb:280]: I think there are no functional changes in this file? all the whitespace got changed

File: .rakeTasks (hunk 1 above)
Comment (cgranleese-r7) [reply .rakeTasks:1]: I was wondering about that, but after reading this, I assumed it was ok.
image

Comment (cgranleese-r7) [reply .rakeTasks:1]: Will get that removed.

**Comment by kernelsmith** [comment]
I don't see why it wouldn't, but you might consider testing this w/the alias plugin as well, something like

- [ ] `load alias`
- [ ] `alias ejecutar run`
- [ ] `use exploit/windows/wins/ms04_045_wins` 
- [ ] type`ejecutar` followed by a space then hit tab and verify as above etc

**Comment by cgranleese-r7** [comment]
> I don't see why it wouldn't, but you might consider testing this w/the alias plugin as well, something like
> 
> * [ ]  `load alias`
> * [ ]  `alias ejecutar run`
> * [ ]  `use exploit/windows/wins/ms04_045_wins`
> * [ ]  type`ejecutar` followed by a space then hit tab and verify as above etc

Tried this but it didn't work. Will raise an issue to tackle it later!

File: lib/msf/ui/console/module_option_tab_completion.rb [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+            if service.host_id == host.id
+              res << service.port.to_s
+            end
+          end
+          return res
+        end
+      end
+    end
+  end
+end
```
Comment (adfoster-r7) [inline lib/msf/ui/console/module_option_tab_completion.rb:350]: Looks like you're missing a trailing new line here 👀 

It's best practice to add a new line to the end of your files, you can do this manually - or you can modify your rubymine settings to do it for you automatically:

image

Comment (adfoster-r7) [reply lib/msf/ui/console/module_option_tab_completion.rb:350]: Context:

https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

Comment (cgranleese-r7) [reply lib/msf/ui/console/module_option_tab_completion.rb:350]: Ok, will get those changes made to my settings and try get this fixed.

**Comment by adfoster-r7** [comment]
Pushed some tweaks to this branch, I've confirmed this as working on Windows + Mac now :+1:

**Comment by dwelch-r7** [comment]
# Release Notes
This PR adds tab completion for run command

**Pull request merged**